packages/cursorless-vscode/src/keyboard/KeyboardCommandsModalLayer.ts

```diff
@@ -0,0 +1,87 @@
+import * as vscode from "vscode";
+import KeyboardHandler from "./KeyboardHandler";
+import TrieSearch from "trie-search";
+import { buildSuffixTrie, KeyValuePair } from "./buildSuffixTrie";
+
+/**
+ * Defines a single keyboard layer to use with a modal version of Cursorless
+ * keyboard. We construct one of these every time the parser changes state,
+ * based on the allowable tokens at the given state.
+ */
+export class KeyboardCommandsModalLayer<Param extends { type: any }> {
+  /**
+   * Merged map from all the different sections of the key map (eg actions,
+   * colors, etc).
+   */
+  private trie: TrieSearch<{
+    key: string;
+    value: Param;
+  }>;
+
+  constructor(
+    private keyboardHandler: KeyboardHandler,
+    entries: [string, Param][],
+  ) {
+    this.handleInput = this.handleInput.bind(this);
+    if (entries.length === 0) {
+      vscode.window.showErrorMessage("No keybindings found for current layer");
+    }
+    const { trie, conflicts } = buildSuffixTrie(entries);
+
+    for (const conflict of conflicts) {
+      const conflictStr = conflict
+        .map(({ key, value: { type } }) => `\`${type}.${key}\``)
+        .join(" and ");
+      vscode.window.showErrorMessage(`Conflicting keybindings: ${conflictStr}`);
+    }
+
+    this.trie = trie;
+  }
+
+  async handleInput(text: string, { previousKeys }: { previousKeys: string }) {
+    let values: KeyValuePair<Param>[];
+
+    if (text === "") {
+      values = [];
+    } else {
+      values = this.trie.search(text);
+
+      if (values.length === 0) {
+        // If we haven't consumed any input yet, then it means the first
+        // character was a false start so we should cancel the whole thing.
+        const errorMessage = `Invalid key '${text}'`;
+        vscode.window.showErrorMessage(errorMessage);
+        throw Error(errorMessage);
+      }
+    }
+
+    let sequence = text;
+
+    // We handle multi-key sequences by repeatedly awaiting a single keypress
+    // until they've pressed something in the map.
+    while (values.length !== 1) {
```

Comment:
Note that checking this way means that not only can the user press any unique suffix, they can press any unique prefix. So you can either structure your maps to use leading differentiator or trailing differentiator and get the same benefit, eg:

leading:

```json
{
  "sf": "namedFunction",
  "sl": "line",
  "st": "token",
}
```

or trailing:

```json
{
  "fs": "namedFunction",
  "ls": "line",
  "ts": "token",
}
```

In either case, if the only thing the current layer is expecting is a scope, then these will shorten to `f`, `l`, and `t` in either case

Reply:
I think(?) that’s good.

The only potential downside I see is that it is hard to mentally model, so minor changes in your bindings can impact other distant bindings.

But let’s just try it out and see. If that proves to be a problem, we have options: Can change approach, or build tooling to generate all possible keymaps, which can be inspected and diffed.

Reply:
Yeah, tho I think the suffix behaviour has the exact same problem, no?

Reply:
Also, generating all possible keymaps isn't super easy, because it requires enumerating all possible paths through the parser, which I believe could be a challenge as Earley parsers support nondeterministic CFGs

Reply:
/me waves hands and adopts Yiddish accent

enough worrying already, you’ll worry your mother to her grave. too many confused users? You should be so lucky.